src/irmin-pack/unix/ext.ml

```diff
@@ -702,20 +702,38 @@ module Maker (Config : Conf.S) = struct
         with exn -> catch_errors "Check GC" exn
 
       let wait repo =
-        try
-          let* _ = finalise_exn ~wait:true repo in
-          Lwt.return_ok ()
-        with exn -> catch_errors "Wait for GC" exn
+        (* Use a loop to check status to avoid issues with
+           multiple threads calling the underlying blocking
+           implementation. {!run} has the only thread that
+           is allowed to block. *)
+        let rec check () =
+          let* is_finished = is_finished repo in
+          match is_finished with
+          | Ok true -> Lwt.return_ok ()
+          | Ok false ->
+              let* () = Lwt_unix.sleep 0.1 in
```

Comment:
I'm not sure if this sleep is necessary but seemed like a good idea to avoid a tight polling loop. Open to thoughts and suggestions!